posts/inside-rust/2020-02-26-ffi-unwind-design-meeting.md

```diff
@@ -0,0 +1,265 @@
+  * With `panic=abort`: all types of unwinding abort the program
+
+This proposal provides 2 ABIs, each suited for different purposes: you would
+generally use `extern "C"` when interacting with C APIs (making sure to avoid
+destructors where `longjmp` might be used), and `extern "C unwind"`
+when interacting with C++ APIs. The main advantage of this proposal is that
+switching between `panic=unwind` and `panic=abort` does not introduce UB if you
+have correctly marked all potential unwinding calls as `"C unwind"` (your
+program will abort instead).
+
+### Proposal 2: Introduce `"C unwind"`, forced unwinding always permitted
+
+This is the same as the previous design, except that when compiled with
+`panic=abort`, forced unwinding would *not* be intercepted at `"C unwind"` ABI
+boundaries; that is, they would behave normally (though still UB if there are
+any destructors), without causing the program to abort. `panic`-unwind and
+non-forced foreign exceptions would still cause the program to abort.
+
+The advantage of treating forced unwinding differently is that it reduces
+portability incompatibilities. Specifically, it ensures that using `"C unwind"`
+cannot cause `longjmp` or `pthread_exit` to stop working (abort the program)
+when the target platform and/or compile flags are changed.  With proposal 1,
+`longjmp` will be able to cross `"C unwind"` boundaries _except_ on Windows
+with MSVC under `panic=abort`, and `pthread_exit` will work inside `"C unwind"`
+functions _except_ when linked with glibc under `panic=abort`. The downside of
+this proposal is that the abort stubs around `"C unwind"` calls in `panic=abort`
+become more complicated since they need to distinguish between different types
+of foreign exceptions.
+
+### Proposal 3: No new ABI
+
+* `panic=unwind`: unwind behaves normally
+* `panic=abort`:
+  * `panic`-unwind: does not exist; `panic!` aborts the program
+  * forced unwinding, no destructors: unwind behaves normally
+  * other foreign unwinding: undefined behavior
+
+The main advantage of this proposal is its simplicity: there is only one ABI and
+the behavior of `panic=abort` is identical to that of `-fno-exceptions` in C++.
+However this comes with the downside that switching to `panic=abort` may in some
+cases introduce UB (though only in unsafe code) if FFI calls unwind through Rust
+code.
+
+Another advantage is that forced unwinding from existing functions defined in
+the `libc` crate such as `pthread_exit` and `longjmp` will be able to unwind
+frames with destructors when compiled with `panic=unwind`, which is not possible
+with the other proposals.
+
+## Comparison table for the proposed designs
+
+In this table, "UB" stands for "undefined behavior". Some instances of UB can
+be detected at runtime, but the code to do so would impose an undesirable
+code-size penalty; the group recommends generating such code only for debug
+builds. These cases are marked "UB (debug: abort)"
```

Comment:
We don't have any cases marked "UB (debug: abort)" in the table.

Reply:
It seems to me that if we didn't care too much about code size for debug builds, we could make all the UB entries be "debug: abort". What do you think?

Reply:
...well, no, I guess that's not true for cases where forced unwinding is allowed but destructors won't be invoked.

Reply:
So would it be possible to use abort shims for all of the UB scenarios in the last column, "other foreign unwind"?

Reply:
We can detect UB in the last column with abort shims around all FFI calls. We can detect UB from forced unwind running destructors by still emitting destructor landing pads but making them abort. In both cases we're pretty much doing codegen as if using `panic=unwind`, which might be too much of a cost even for debug builds.

Reply:
In the second-to-last column, though, destructors _won't_ reliably be run, even in debug mode, right?

Reply:
Actually, I guess there's no reason why they wouldn't.

Reply:
Since, as you say, we'd be doing codegen as if it were `panic=unwind`